optax/_src/transform.py

```diff
@@ -869,6 +869,64 @@ def update_fn(updates, state, params=None):
   return base.GradientTransformation(init_fn, update_fn)
 
 
+class ScaleByNewtonState(NamedTuple):
+  """State holding the inverse of the sum of gradient squares to date."""
+
+  hessian_inv: base.Updates
+
+
+def sherman_morrison(a_inv, u, v):
+  den = 1.0 + (u @ a_inv) @ v
+  a_inv -= a_inv @ jnp.outer(u, v) @ a_inv / den
+  return a_inv
+
+
+def scale_by_newton(eps: float = 1e-7) -> base.GradientTransformation:
+  """Rescale updates by multiplying by the inverse of an approximation of the
+  hessian.
+
+  References
+  ----------
+  [^1]: [Hazan, E., Agarwal, A. and Kale, S., 2007. Logarithmic regret
+  algorithms for online convex optimization. Machine Learning, 69(2-3),
+  pp.169-192](https://www.cs.princeton.edu/~ehazan/papers/log-journal.pdf)
+
+  Args:
+    eps: A small floating point value to avoid zero denominator.
+
+  Returns:
+    An (init_fn, update_fn) tuple.
+  """
+
+  def init_fn(params):
+    hessian_inv = jax.tree_map(
+      lambda t: jnp.eye(len(t.flatten()), dtype=t.dtype) / eps, params
+    )
+    return ScaleByNewtonState(hessian_inv=hessian_inv)
+
+  def update_fn(updates, state, params=None):
+    del params
+
+    class Tuple(tuple):
```

Comment:
Do I understand correctly that you wouldn't need this if you don't overwrite `updates` in line 917 and instead introduce a new variable e.g. `flattened_updates` and then in line 922:

```
updates = jax.tree_multimap(
  lambda flat_u, u: flat_u.reshape(u.shape), flattened_updates, updates)
```

If that's true I think it would be good to simplify the code and remove the definition the `Tuple`.

Reply:
@eserie

Reply:
Thank you for the observation! I was a bit messy in my implementation, your proposal seems much more reasonable. I updated the code accordingly and also add a test to cover the multi-dimensional case.